docs/src/main/asciidoc/getting-started-dev-services.adoc

```diff
@@ -0,0 +1,326 @@
+
+== Setting up an interactive application
+
+=== Bootstrapping the project
+
+The easiest way to create a new Quarkus project is to open a terminal and run the following command:
+
+For Linux & MacOS users
+
+:create-app-artifact-id: getting-started-dev-services
+:create-app-extensions: resteasy-reactive
+:create-app-code:
+include::{includes}/devtools/create-app.adoc[]
+
+For Windows users:
+
+- If using cmd , (don't use backward slash `\` and put everything on the same line)
+- If using Powershell , wrap `-D` parameters in double quotes e.g. `"-DprojectArtifactId=getting-started"`
+
+For an explanation of what's in the generated application, see the xref:getting-started.adoc[First application guide].
+
+=== Running the application
+
+Launch the application in dev mode
+----
+mvn quarkus:dev
+----
+
+Once the application is up, visit http://localhost:8080/hello. It should show a "Hello from RestEasy Reactive" message.
+
+=== Accepting user input
+
+Let's make the application a bit more interactive.
+Open the project in your IDE and navigate to `src/main/java/org/acme/GreetingResource.java'
+Add a query param in the `hello` method.
+
+----
+public String hello(@QueryParam("name") String name) {
+return "Hello " + name;
+}
+----
+
+Visit http://localhost:8080/hello?name=Bloom.
+
+You should see a personalised message: `Hello Bloom`.
+
+=== Fixing the tests
+
+In your Quarkus terminal, type 'r' to run the tests. You should see
+that your application changes broke the tests!
+
+To fix the tests, open `src/test/java/org/acme/GreetingResource.test`
+and replace
+
+----
+             .body(is("Hello from RESTEasy Reactive"));
+----
+
+with
+
+----
+             .body(containsString("Hello"));
+----
+
+This still validates the rest endpoint, but it's more flexible
+about the expected output.
+You should see in your terminal that the tests are now passing.
+
+== Adding persistence
+
+=== Creating a Panache Entity
+
+1. To add the persistence libraries, run
+----
+quarkus ext add hibernate-orm-panache jdbc-postgresql
+----
+
+The application will record the names of people it greets. Define a simple Hibernate Entity
+by creating a `Greeting.java` class. Add the following content:
+
+----
+@Entity
+public class Greeting extends PanacheEntity {
+String name;
+}
+----
+
+The entity makes use of xref:{doc-guides}hibernate-orm-panache.adoc[Panache], a layer on top of Hibernate.
+ Extending `PanacheEntity` brings in a range of methods for reading, writing, and finding data.
+ Because all the data access methods are on the `Greeting` entity, rather than on a separate data access class,
+ this is an example of the active record pattern.
+
+The `Greeting` table will have one column, a field called `name`.
+
+=== Writing data
+
+To use the new entity, update the `hello` method to start writing some data.
+
+Change the method to the following:
+
+----
+@GET
+@Transactional
+@Produces(MediaType.TEXT_PLAIN)
+public String hello(@QueryParam("name") String name) {
```

Comment:
This is slightly annoying...

As GET must be idempotent, and so should not change the server state.
I would recommend using @POST (or having an @POST endpoint).

Reply:
I knooow, but if we use post, it's (a) harder to test and (b) less realistic. No one would expect to do a post to retrieve information, if the information under discussion isn't changed by the request. Many pages store analytics about site traffic to their @GET endpoints. Since something somewhere changed, arguably the GET endpoint is changing state. No website ever has said 'we use Google Analytics, so you must use POST to visit our pages.' :)

Reply:
I guess it's a question of the semantics of what the server state is. It's sort of hard to reason about semantics for an endpoint that just says "hello," because the whole premise doesn't actually have value. But I would argue the 'intended effect' is to say hello, in an idempotent way. The persistence of the name is a traffic-monitoring side effect.